MantidPlot/src/ProjectRecoveryGUIs/ProjectRecoveryPresenter.cpp

```diff
@@ -0,0 +1,190 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "ProjectRecoveryPresenter.h"
+#include "ApplicationWindow.h"
+#include "ProjectRecovery.h"
+#include "ProjectRecoveryModel.h"
+#include "ProjectRecoveryView.h"
+#include "RecoveryFailureView.h"
+
+#include <QDialog>
+#include <memory>
+
+ProjectRecoveryPresenter::ProjectRecoveryPresenter(
+    MantidQt::ProjectRecovery *projectRecovery, ApplicationWindow *parentWindow)
+    : m_mainWindow(parentWindow) {
+  m_recView = nullptr;
+  m_failureView = nullptr;
+  m_model = new ProjectRecoveryModel(projectRecovery, this);
+  m_openView = RecoveryView;
+  m_startMantidNormallyCalled = false;
+}
+
+ProjectRecoveryPresenter::ProjectRecoveryPresenter(
+    const ProjectRecoveryPresenter &obj) {
+  /// Copy constructor can only copy the Model
+  m_recView = obj.m_recView;
+  m_failureView = obj.m_failureView;
+  m_model = new ProjectRecoveryModel(nullptr, this);
+  *m_model = *obj.m_model;
+  m_mainWindow = obj.m_mainWindow;
+  m_openView = obj.m_openView;
+  m_startMantidNormallyCalled = obj.m_startMantidNormallyCalled;
+}
+
+ProjectRecoveryPresenter::~ProjectRecoveryPresenter() {
+  delete m_recView;
+  delete m_failureView;
+  delete m_model;
+}
+
+bool ProjectRecoveryPresenter::startRecoveryView() {
+  try {
+    m_recView = new ProjectRecoveryView(m_mainWindow, this);
+    m_openView = RecoveryView;
+    m_recView->exec();
+  } catch (...) {
+    return true;
+  }
+
+  // If start mantid normally was called we want to cancel
+  if (m_startMantidNormallyCalled) {
+    return false;
+  }
+
+  // If run has failed and recovery is not running
+  if (m_model->getFailedRun()) {
+    return true;
+  }
+  return false;
+}
+
+bool ProjectRecoveryPresenter::startRecoveryFailure() {
+  try {
+    m_failureView = new RecoveryFailureView(m_mainWindow, this);
+    m_openView = FailureView;
+    m_failureView->exec();
+  } catch (...) {
+    return true;
+  }
+
+  // If start mantid normally was called we want to cancel
+  if (m_startMantidNormallyCalled) {
+    return false;
+  }
+
+  // If run has failed and recovery is not running
+  if (m_model->getFailedRun()) {
+    return true;
+  }
+  return false;
+}
+
+QStringList ProjectRecoveryPresenter::getRow(int i) {
+  auto vec = m_model->getRow(i);
+  QStringList returnVal;
+  for (auto i = 0; i < 3; ++i) {
```

Comment:
Could the magic number 3 have a descriptive name?

Reply:
Replaced with what it actively represents :+1: